docs/composability.md

```diff
@@ -0,0 +1,24 @@
+
+
+
+# Making the model "pipeline friendly"
+Whether you plan to manually edit your model code, or use techniques like tracing to extract model chunks, you will have to construct nn.Module objects representing the portion of the model that runs on a given pipeline stage.  Regardless of how you approach this, a few changes to the original model code can go a long way to making this process easier.
+
+### Simplify the top-level model forward
+Most likely, you can write your model in such a way that the top-level nn.Module owns a sequence of child modules that it calls during forward, delegating most of the complexity to the child module forwards.  If you can reduce your top level forward to mostly a for-loop over child module calls, then you'll simplify the pipeline-partitioning task to choosing the set of submodules to keep per stage.  If you have non-trivial logic in the top-level forward, you'll have to find a way to patch that logic back onto the resulting pipeline stage model, which can be annoying.
+
+example ([PR #321](https://github.com/pytorch/torchtitan/pull/321)):
+we used to slice the freqs_cis buffer by seq_len in the top level forward, pass that into child modules, and expect that inside the child modules the seq_len would match up with the size of other local tensors.  But we don't know about whether TP was applied or not when we consider PP splitting and could create a mismatch.  Its just as easy to perform the freqs_cis slicing inside the child submodule, using the runtime-accurate local seq_len, and this sidesteps the issue at PP slicing time.  
+
+example ([PR #322])https://github.com/pytorch/torchtitan/pull/322)): We decided to actually reuse the top-level model object on every PP stage, just delete the layers we don't want, and make sure that the top-level forward would do the right thing.  This means we don't have to make a separate runtime pp_forward that glues together child modules per stage.  The first change was using a moduledict instead of modulelist to store layers preserves layer FQNs even when deleting - e.g. layers.1 stays layers.1 even if you remove layers.0, which isn't true for a list- this matters for checkpoint save/load.  The second change was making the input and output layers optional- if the layer exists, we run it, otherwise we feed the input through to bypass it.  With these two changes, we can just (meta)-initialize the whole model, delete the unused parts per stage, then materialize the remaining part on GPU before loading a checkpoint.
```

Comment:
Should we explain the unique FQNs is the requirement of using DCP? People may be more familiar with using rank as the file name with torch.save, which does not require the unique FQNs.

Reply:
Yeah I agree, let's clarify this is a requirement of DCP to save/load sharded state dict